File: bmd_emulator/protocol.py
"""Blackmagic Streaming Encoder Ethernet Protocol server.

Implements enough of the v1.2 protocol to look like a real Streaming
Encoder to clients (the BMD Streaming Setup utility, ATEM software,
or any custom integration). Listens on TCP 9977.

Protocol shape:
  - Line-oriented, ASCII.
  - Server lines end with LF; client lines may end with LF or CRLF.
  - Messages are blocks: a header line ending in `:`, then key:value
    lines, then a blank line.
  - On connect, the server dumps a preamble + every status block, then
    `END PRELUDE:` + blank line.
  - Client sends a block to mutate state; server replies `ACK` or
    `NACK`, then a fresh status block reflecting the change.
"""

from __future__ import annotations

import logging
import socket
import threading
from typing import Callable

from .state import (
    AVAILABLE_VIDEO_MODES,
    EncoderState,
    QUALITY_LEVELS,
)
from .streamer import Streamer

log = logging.getLogger(__name__)

PROTOCOL_VERSION = "1.2"
PORT = 9977


class ProtocolServer:
    def __init__(self, state: EncoderState, streamer: Streamer, port: int = PORT):
        self.state = state
        self.streamer = streamer
        self.port = port
        self._sock: socket.socket | None = None
        self._thread: threading.Thread | None = None
        self._clients: list[socket.socket] = []
        self._clients_lock = threading.Lock()
        self._stop = threading.Event()
        self._pending_block: dict[socket.socket, tuple[str, list[tuple[str, str]]]] = {}

    # ---- lifecycle ---------------------------------------------------------

    def start(self) -> None:
        if self._thread is not None:
            return
        # Port-walk fallback: try the requested port first (9977 is the
        # documented BMD port), then up to nine spots forward. A real
        # BMD encoder on the LAN doesn't conflict (we bind 0.0.0.0
        # locally; the LAN device has its own bind), but a stale
        # instance on the same machine would silently brick a fresh
        # launch otherwise.
        last_err: OSError | None = None
        for offset in range(10):
            try_port = self.port + offset
            sock = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
            sock.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
            try:
                sock.bind(("0.0.0.0", try_port))
                self._sock = sock
                self.port = try_port
                last_err = None
                break
            except OSError as exc:
                sock.close()
                last_err = exc
        if self._sock is None:
            raise RuntimeError(
                f"Could not bind BMD protocol on TCP {self.port}-{self.port + 9}: {last_err}"
            )
        self._sock.listen(8)
        self._sock.settimeout(0.5)
        self._stop.clear()
        self._thread = threading.Thread(
            target=self._accept_loop, name="bmd-protocol", daemon=True
        )
        self._thread.start()
        log.info("BMD control protocol listening on TCP %d", self.port)

    def stop(self) -> None:
        self._stop.set()
        if self._sock is not None:
            try:
                self._sock.close()
            except OSError:
                pass
        with self._clients_lock:
            for c in self._clients:
                try:
                    c.close()
                except OSError:
                    pass
            self._clients.clear()

    # ---- accept loop -------------------------------------------------------

    def _accept_loop(self) -> None:
        assert self._sock is not None
        while not self._stop.is_set():
            try:
                client, addr = self._sock.accept()
            except socket.timeout:
                continue
            except OSError:
                break
            log.info("control client connected: %s", addr)
            with self._clients_lock:
                self._clients.append(client)
            t = threading.Thread(
                target=self._client_loop,
                args=(client, addr),
                daemon=True,
                name=f"bmd-client-{addr[1]}",
            )
            t.start()

    def _client_loop(self, client: socket.socket, addr) -> None:
        try:
            self._send_preamble(client)
            buf = b""
            while not self._stop.is_set():
                try:
                    chunk = client.recv(4096)
                except OSError:
                    break
                if not chunk:
                    break
                buf += chunk
                while b"\n" in buf:
                    line, buf = buf.split(b"\n", 1)
                    self._feed_line(client, line.decode("utf-8", "replace"))
        finally:
            with self._clients_lock:
                if client in self._clients:
                    self._clients.remove(client)
            try:
                client.close()
            except OSError:
                pass
            log.info("control client disconnected: %s", addr)

    # ---- block I/O ---------------------------------------------------------

    def _send(self, sock: socket.socket, lines: list[str]) -> None:
        payload = ("\n".join(lines) + "\n").encode("utf-8")
        try:
            sock.sendall(payload)
        except OSError:
            pass

    def _send_block(self, sock: socket.socket, header: str, kvs: list[tuple[str, str]]) -> None:
        lines = [f"{header}:"]
        for k, v in kvs:
            lines.append(f"{k}: {v}")
        lines.append("")
        self._send(sock, lines)

    def _broadcast(self, header: str, kvs: list[tuple[str, str]]) -> None:
        with self._clients_lock:
            clients = list(self._clients)
        for c in clients:
            self._send_block(c, header, kvs)

    # ---- preamble + dump ---------------------------------------------------

    def _send_preamble(self, sock: socket.socket) -> None:
        self._send_block(sock, "PROTOCOL PREAMBLE", [("Version", PROTOCOL_VERSION)])
        self._send_block(sock, "IDENTITY", self._identity_kvs())
        self._send_block(sock, "VERSION", self._version_kvs())
        for line_group in self._network_lines():
            self._send(sock, line_group)
        self._send_block(sock, "UI SETTINGS", self._ui_kvs())
        self._send_block(sock, "STREAM SETTINGS", self._stream_settings_kvs())
        self._send_block(sock, "STREAM XML", [("Files", ", ".join(self.state.services.keys()))])
        self._send_block(sock, "STREAM STATE", self._stream_state_kvs())
        self._send_block(sock, "AUDIO SETTINGS", self._audio_kvs())
        self._send_block(sock, "END PRELUDE", [])

    # ---- block builders ----------------------------------------------------

    def _identity_kvs(self) -> list[tuple[str, str]]:
        s = self.state
        return [
            ("Model", s.model),
            ("Label", s.label),
            ("Unique ID", s.unique_id),
        ]

    def _version_kvs(self) -> list[tuple[str, str]]:
        return [
            ("Product ID", "BE73"),
            ("Hardware Version", "0100"),
            ("Software Version", "01000000"),
            ("Software Release", "0.1"),
        ]

    def _network_lines(self) -> list[list[str]]:
        block: list[str] = ["NETWORK:", "Interface Count: 1", "Default Interface: 0", ""]
        iface: list[str] = [
            "NETWORK INTERFACE 0:",
            "Name: Ethernet",
            "Priority: 1",
            "MAC Address: 00:11:22:33:44:55",
            "Dynamic IP: true",
            "Current Addresses: 0.0.0.0/255.255.255.0",
            "Current Gateway: 0.0.0.0",
            "Current DNS Servers: ",
            "Static Addresses: 0.0.0.0/255.255.255.0",
            "Static Gateway: 0.0.0.0",
            "Static DNS Servers: 8.8.8.8, 8.8.4.4",
            "",
        ]
        return [block, iface]

    def _ui_kvs(self) -> list[tuple[str, str]]:
        return [
            ("Available Locales", "en_US.UTF-8"),
            ("Current Locale", "en_US.UTF-8"),
            ("Available Audio Meters", "PPM -18dB, PPM -20dB, VU -18dB, VU -20dB"),
            ("Current Audio Meter", "PPM -20dB"),
        ]

    def _stream_settings_kvs(self) -> list[tuple[str, str]]:
        s = self.state
        snap = s.snapshot()
        url = snap.get("current_url", "") or ""
        servers = snap.get("available_servers") or []
        server_names = ", ".join(sv["name"] for sv in servers) or "SRT"
        return [
            ("Available Video Modes", ", ".join(AVAILABLE_VIDEO_MODES)),
            ("Video Mode", s.video_mode),
            ("Current Platform", s.current_service_name or "My Platform"),
            ("Current Server", s.current_server_name),
            ("Current Quality Level", s.quality_level),
            ("Stream Key", s.stream_key),
            ("Password", s.passphrase),
            ("Current URL", url),
            ("Customizable URL", "true"),
            ("Available Default Platforms", ""),
            ("Available Custom Platforms", ", ".join(self.state.services.keys())),
            ("Available Servers", server_names),
            ("Available Quality Levels", ", ".join(QUALITY_LEVELS)),
        ]

    def _stream_state_kvs(self) -> list[tuple[str, str]]:
        s = self.state.stats
        return [
            ("Status", s.status),
            ("Bitrate", str(s.bitrate)),
            ("Duration", s.duration_string()),
            ("Cache Used", str(s.cache_used_pct)),
        ]

    def _audio_kvs(self) -> list[tuple[str, str]]:
        return [
            ("Current Monitor Out Audio Source", "Auto"),
            ("Available Monitor Out Audio Sources", "Auto, SDI In, Remote Source"),
        ]

    # ---- line accumulator → block dispatcher -------------------------------

    def _feed_line(self, client: socket.socket, line: str) -> None:
        line = line.rstrip("\r")
        pending = self._pending_block.get(client)
        if pending is None:
            if not line.strip():
                return
            if line.endswith(":"):
                header = line[:-1].strip()
                self._pending_block[client] = (header, [])
            return

        if line.strip() == "":
            header, kvs = pending
            del self._pending_block[client]
            self._dispatch(client, header, kvs)
            return

        if ":" in line:
            k, v = line.split(":", 1)
            pending[1].append((k.strip(), v.strip()))

    # ---- dispatch ----------------------------------------------------------

    def _dispatch(self, client: socket.socket, header: str, kvs: list[tuple[str, str]]) -> None:
        log.info("RX block %r %r", header, kvs)
        handler: Callable[[socket.socket, list[tuple[str, str]]], bool] | None = {
            "IDENTITY": self._handle_identity,
            "STREAM SETTINGS": self._handle_stream_settings,
            "STREAM STATE": self._handle_stream_state,
            "STREAM XML": self._handle_stream_xml,
            "NETWORK INTERFACE 0": self._handle_network_iface,
            "UI SETTINGS": self._handle_ui_settings,
            "AUDIO SETTINGS": self._handle_audio,
            "SHUTDOWN": self._handle_shutdown,
        }.get(header)

        if handler is None:
            self._send_block(client, "NACK", [])
            return

        try:
            ok = handler(client, kvs)
        except Exception as exc:  # noqa: BLE001
            log.exception("handler failed")
            self._send_block(client, "NACK", [])
            self._send_block(client, "ERROR", [("Message", str(exc))])
            return

        if ok:
            self._send_block(client, "ACK", [])
            # Re-dump the affected block to all clients so they stay in sync.
            self._broadcast_after(header)
        else:
            self._send_block(client, "NACK", [])

    def _broadcast_after(self, header: str) -> None:
        if header == "STREAM SETTINGS" or not header:
            self._broadcast("STREAM SETTINGS", self._stream_settings_kvs())
        elif header == "STREAM STATE":
            self._broadcast("STREAM STATE", self._stream_state_kvs())
        elif header == "IDENTITY":
            self._broadcast("IDENTITY", self._identity_kvs())
        elif header == "AUDIO SETTINGS":
            self._broadcast("AUDIO SETTINGS", self._audio_kvs())

    # ---- handlers ----------------------------------------------------------

    def _handle_identity(self, client, kvs) -> bool:
        for k, v in kvs:
            if k == "Label":
                self.state.label = v
        return True

    def _handle_stream_settings(self, client, kvs) -> bool:
        s = self.state
        with s.lock:
            for k, v in kvs:
                if k == "Video Mode" and v in AVAILABLE_VIDEO_MODES:
                    s.video_mode = v
                elif k == "Current Platform":
                    if v in s.services:
                        s.current_service_name = v
                        s.stream_key = s.services[v].key
                elif k == "Current Quality Level":
                    s.quality_level = v
                elif k == "Stream Key":
                    s.stream_key = v
                elif k == "Password":
                    s.passphrase = v
                elif k == "Current Server":
                    s.current_server_name = v
        return True

    def _handle_stream_state(self, client, kvs) -> bool:
        for k, v in kvs:
            if k == "Action":
                action = v.strip().lower()
                if action == "start":
                    try:
                        self.streamer.start()
                    except Exception as exc:  # noqa: BLE001
                        self.state.stats.status = "Interrupted"
                        self.state.stats.error = str(exc)
                        return False
                elif action == "stop":
                    self.streamer.stop()
        return True

    def _handle_stream_xml(self, client, kvs) -> bool:
        # We accept STREAM XML <name>: blocks for adding a service, but
        # the body is the full XML document — we don't inline-parse it
        # in this emulator. Adding files happens via the HTTP UI instead.
        return True

    def _handle_network_iface(self, client, kvs) -> bool:
        # Read-only here; pretend we accepted it for compatibility.
        return True

    def _handle_ui_settings(self, client, kvs) -> bool:
        return True

    def _handle_audio(self, client, kvs) -> bool:
        return True

    def _handle_shutdown(self, client, kvs) -> bool:
        for k, v in kvs:
            if k == "Action" and v.strip().lower() == "factory reset":
                self.streamer.stop()
                self.state.stats.error = None
                return True
        return True

File: src-tauri/src/protocol.rs
//! Blackmagic Streaming Encoder Ethernet Protocol — TCP 9977.
//!
//! Implements enough of the v1.2 protocol to look like a real
//! Streaming Encoder to BMD's Streaming Setup utility, ATEM software,
//! or any custom integration. Line-oriented ASCII; messages are
//! "blocks" (header line ending `:`, key/value lines, blank line).
//!
//! Boot order: server connects -> sends a preamble (PROTOCOL PREAMBLE
//! / IDENTITY / VERSION / NETWORK / UI SETTINGS / STREAM SETTINGS /
//! STREAM XML / STREAM STATE / AUDIO SETTINGS / END PRELUDE) -> reads
//! client blocks one at a time. For each client block the server
//! either ACKs and re-broadcasts the affected status to all clients,
//! or NACKs.
//!
//! Architecture: one tokio task per accepted client. Broadcasts use
//! a tokio::sync::broadcast channel that every client task subscribes
//! to. The server itself owns the listener on a free port (9977
//! preferred, walks 9977..9986 to dodge stale instances on the same
//! machine).

use crate::state::{EncoderState, AVAILABLE_VIDEO_MODES};
use crate::streamer::Streamer;

use anyhow::{anyhow, Result};
use std::net::SocketAddr;
use std::sync::Arc;
use tokio::io::{AsyncBufReadExt, AsyncWriteExt, BufReader};
use tokio::net::{TcpListener, TcpStream};
use tokio::sync::{broadcast, Mutex};

const PROTOCOL_VERSION: &str = "1.2";
const QUALITY_LEVELS: &[&str] = &["Streaming High", "Streaming Medium", "Streaming Low"];

#[derive(Clone, Debug)]
struct BroadcastBlock {
    header: String,
    kvs: Vec<(String, String)>,
}

pub struct ProtocolServer {
    state: Arc<EncoderState>,
    streamer: Arc<Streamer>,
    broadcast: broadcast::Sender<BroadcastBlock>,
}

impl ProtocolServer {
    pub fn new(state: Arc<EncoderState>, streamer: Arc<Streamer>) -> Arc<Self> {
        let (tx, _) = broadcast::channel(64);
        Arc::new(Self {
            state,
            streamer,
            broadcast: tx,
        })
    }

    /// Bind the BMD protocol on 9977 (or the next free port up to
    /// 9986) and start the accept loop. Returns the port we actually
    /// bound — the dev UI shows it in the about-this-instance label.
    pub async fn start(self: &Arc<Self>, start_port: u16) -> Result<u16> {
        let (port, listener) = bind_with_walk(start_port).await?;
        log::info!("BMD control protocol listening on TCP {port}");

        let me = self.clone();
        tokio::spawn(async move {
            loop {
                match listener.accept().await {
                    Ok((stream, addr)) => {
                        log::info!("BMD control client connected: {addr}");
                        let me = me.clone();
                        tokio::spawn(async move {
                            if let Err(err) = me.client_loop(stream, addr).await {
                                log::warn!("BMD client loop {addr} ended: {err}");
                            } else {
                                log::info!("BMD control client disconnected: {addr}");
                            }
                        });
                    }
                    Err(err) => {
                        log::warn!("BMD accept failed: {err}");
                        // Brief backoff so a broken socket doesn't
                        // burn a CPU core in the loop.
                        tokio::time::sleep(std::time::Duration::from_millis(100)).await;
                    }
                }
            }
        });
        Ok(port)
    }

    async fn client_loop(self: Arc<Self>, stream: TcpStream, _addr: SocketAddr) -> Result<()> {
        stream.set_nodelay(true)?;
        let (read, write) = stream.into_split();
        let write = Arc::new(Mutex::new(write));

        // Initial dump (preamble + every block + END PRELUDE).
        self.send_preamble(write.clone()).await?;

        // Subscribe to subsequent broadcasts and forward them on a
        // separate task so the read loop never blocks.
        let mut rx = self.broadcast.subscribe();
        let write_for_bcast = write.clone();
        let bcast_task = tokio::spawn(async move {
            while let Ok(msg) = rx.recv().await {
                if send_block(&write_for_bcast, &msg.header, &msg.kvs).await.is_err() {
                    break;
                }
            }
        });

        // Read loop — blocks accumulate until a blank line, then
        // dispatch.
        let mut reader = BufReader::new(read).lines();
        let mut pending: Option<(String, Vec<(String, String)>)> = None;
        while let Ok(Some(raw_line)) = reader.next_line().await {
            let line = raw_line.trim_end_matches('\r').to_string();
            match pending.as_mut() {
                None => {
                    if line.trim().is_empty() {
                        continue;
                    }
                    if let Some(header) = line.strip_suffix(':') {
                        pending = Some((header.trim().to_string(), Vec::new()));
                    }
                    // Lines that don't end in `:` and don't have a
                    // pending block are stray — ignore.
                }
                Some((header, kvs)) => {
                    if line.trim().is_empty() {
                        let header = std::mem::take(header);
                        let kvs = std::mem::take(kvs);
                        pending = None;
                        self.dispatch(&write, &header, &kvs).await?;
                        continue;
                    }
                    if let Some((k, v)) = line.split_once(':') {
                        kvs.push((k.trim().to_string(), v.trim().to_string()));
                    }
                }
            }
        }

        bcast_task.abort();
        Ok(())
    }

    async fn dispatch(
        &self,
        write: &Arc<Mutex<tokio::net::tcp::OwnedWriteHalf>>,
        header: &str,
        kvs: &[(String, String)],
    ) -> Result<()> {
        log::info!("BMD RX block {header:?} {kvs:?}");
        let result = match header {
            "IDENTITY" => self.handle_identity(kvs),
            "STREAM SETTINGS" => self.handle_stream_settings(kvs),
            "STREAM STATE" => self.handle_stream_state(kvs).await,
            "STREAM XML" | "NETWORK INTERFACE 0" | "UI SETTINGS" | "AUDIO SETTINGS" => Ok(()),
            "SHUTDOWN" => self.handle_shutdown(kvs).await,
            _ => Err(anyhow!("unknown block header")),
        };

        match result {
            Ok(()) => {
                send_block(write, "ACK", &[]).await?;
                self.broadcast_after(header);
            }
            Err(err) => {
                send_block(write, "NACK", &[]).await?;
                send_block(write, "ERROR", &[("Message".into(), err.to_string())]).await?;
            }
        }
        Ok(())
    }

    fn broadcast_after(&self, header: &str) {
        let snap = self.state.snapshot();
        let msg = match header {
            "STREAM SETTINGS" => BroadcastBlock {
                header: "STREAM SETTINGS".into(),
                kvs: stream_settings_kvs(&snap),
            },
            "STREAM STATE" => BroadcastBlock {
                header: "STREAM STATE".into(),
                kvs: stream_state_kvs(&snap),
            },
            "IDENTITY" => BroadcastBlock {
                header: "IDENTITY".into(),
                kvs: identity_kvs(&snap),
            },
            "AUDIO SETTINGS" => BroadcastBlock {
                header: "AUDIO SETTINGS".into(),
                kvs: audio_kvs(),
            },
            _ => return,
        };
        // Errors mean no subscribers — fine, just drop.
        let _ = self.broadcast.send(msg);
    }

    fn handle_identity(&self, kvs: &[(String, String)]) -> Result<()> {
        for (k, v) in kvs {
            if k == "Label" {
                self.state.set_label(v);
            }
        }
        Ok(())
    }

    fn handle_stream_settings(&self, kvs: &[(String, String)]) -> Result<()> {
        let mut update = crate::state::SettingsUpdate::default();
        for (k, v) in kvs {
            match k.as_str() {
                "Video Mode" if AVAILABLE_VIDEO_MODES.contains(&v.as_str()) => {
                    update.video_mode = Some(v.clone())
                }
                "Current Platform" => update.current_service_name = Some(v.clone()),
                "Current Quality Level" => update.quality_level = Some(v.clone()),
                "Stream Key" => update.stream_key = Some(v.clone()),
                "Password" => update.passphrase = Some(v.clone()),
                "Current Server" => update.current_server_name = Some(v.clone()),
                _ => {}
            }
        }
        self.state.apply_settings(&update);
        Ok(())
    }

    async fn handle_stream_state(&self, kvs: &[(String, String)]) -> Result<()> {
        for (k, v) in kvs {
            if k == "Action" {
                match v.trim().to_lowercase().as_str() {
                    "start" => {
                        let streamer = self.streamer.clone();
                        let result = streamer.start().await;
                        if let Err(err) = result {
                            self.state.stats_in_place(|s| {
                                s.status = "Interrupted".into();
                                s.error = Some(err.to_string());
                            });
                            return Err(err);
                        }
                    }
                    "stop" => {
                        let _ = self.streamer.stop().await;
                    }
                    _ => {}
                }
            }
        }
        Ok(())
    }

    async fn handle_shutdown(&self, kvs: &[(String, String)]) -> Result<()> {
        for (k, v) in kvs {
            if k == "Action" && v.trim().eq_ignore_ascii_case("factory reset") {
                let _ = self.streamer.stop().await;
                self.state.stats_in_place(|s| s.error = None);
            }
        }
        Ok(())
    }

    async fn send_preamble(
        &self,
        write: Arc<Mutex<tokio::net::tcp::OwnedWriteHalf>>,
    ) -> Result<()> {
        let snap = self.state.snapshot();
        send_block(
            &write,
            "PROTOCOL PREAMBLE",
            &[("Version".into(), PROTOCOL_VERSION.into())],
        )
        .await?;
        send_block(&write, "IDENTITY", &identity_kvs(&snap)).await?;
        send_block(&write, "VERSION", &version_kvs()).await?;
        send_lines(
            &write,
            &["NETWORK:", "Interface Count: 1", "Default Interface: 0", ""],
        )
        .await?;
        send_lines(
            &write,
            &[
                "NETWORK INTERFACE 0:",
                "Name: Ethernet",
                "Priority: 1",
                "MAC Address: 00:11:22:33:44:55",
                "Dynamic IP: true",
                "Current Addresses: 0.0.0.0/255.255.255.0",
                "Current Gateway: 0.0.0.0",
                "Current DNS Servers: ",
                "Static Addresses: 0.0.0.0/255.255.255.0",
                "Static Gateway: 0.0.0.0",
                "Static DNS Servers: 8.8.8.8, 8.8.4.4",
                "",
            ],
        )
        .await?;
        send_block(&write, "UI SETTINGS", &ui_kvs()).await?;
        send_block(&write, "STREAM SETTINGS", &stream_settings_kvs(&snap)).await?;
        send_block(
            &write,
            "STREAM XML",
            &[("Files".into(), snap.available_services.join(", "))],
        )
        .await?;
        send_block(&write, "STREAM STATE", &stream_state_kvs(&snap)).await?;
        send_block(&write, "AUDIO SETTINGS", &audio_kvs()).await?;
        send_block(&write, "END PRELUDE", &[]).await?;
        Ok(())
    }
}

// ---- block builders --------------------------------------------------------

fn identity_kvs(snap: &crate::state::Snapshot) -> Vec<(String, String)> {
    vec![
        ("Model".into(), snap.model.clone()),
        ("Label".into(), snap.label.clone()),
        ("Unique ID".into(), snap.unique_id.clone()),
    ]
}

fn version_kvs() -> Vec<(String, String)> {
    vec![
        ("Product ID".into(), "BE73".into()),
        ("Hardware Version".into(), "0100".into()),
        ("Software Version".into(), "01000000".into()),
        ("Software Release".into(), "0.2".into()),
    ]
}

fn ui_kvs() -> Vec<(String, String)> {
    vec![
        ("Available Locales".into(), "en_US.UTF-8".into()),
        ("Current Locale".into(), "en_US.UTF-8".into()),
        (
            "Available Audio Meters".into(),
            "PPM -18dB, PPM -20dB, VU -18dB, VU -20dB".into(),
        ),
        ("Current Audio Meter".into(), "PPM -20dB".into()),
    ]
}

fn audio_kvs() -> Vec<(String, String)> {
    vec![
        (
            "Current Monitor Out Audio Source".into(),
            "Auto".into(),
        ),
        (
            "Available Monitor Out Audio Sources".into(),
            "Auto, SDI In, Remote Source".into(),
        ),
    ]
}

fn stream_settings_kvs(snap: &crate::state::Snapshot) -> Vec<(String, String)> {
    let server_names = if snap.available_servers.is_empty() {
        "SRT".to_string()
    } else {
        snap.available_servers
            .iter()
            .map(|s| s.name.clone())
            .collect::<Vec<_>>()
            .join(", ")
    };
    vec![
        (
            "Available Video Modes".into(),
            AVAILABLE_VIDEO_MODES.join(", "),
        ),
        ("Video Mode".into(), snap.video_mode.clone()),
        (
            "Current Platform".into(),
            if snap.current_service_name.is_empty() {
                "My Platform".into()
            } else {
                snap.current_service_name.clone()
            },
        ),
        ("Current Server".into(), snap.current_server_name.clone()),
        ("Current Quality Level".into(), snap.quality_level.clone()),
        ("Stream Key".into(), snap.stream_key.clone()),
        ("Password".into(), snap.passphrase.clone()),
        ("Current URL".into(), snap.current_url.clone()),
        ("Customizable URL".into(), "true".into()),
        ("Available Default Platforms".into(), String::new()),
        (
            "Available Custom Platforms".into(),
            snap.available_services.join(", "),
        ),
        ("Available Servers".into(), server_names),
        ("Available Quality Levels".into(), QUALITY_LEVELS.join(", ")),
    ]
}

fn stream_state_kvs(snap: &crate::state::Snapshot) -> Vec<(String, String)> {
    vec![
        ("Status".into(), snap.stats.status.clone()),
        ("Bitrate".into(), snap.stats.bitrate.to_string()),
        ("Duration".into(), snap.stats.duration.clone()),
        ("Cache Used".into(), snap.stats.cache_used.to_string()),
    ]
}

// ---- low-level send helpers ------------------------------------------------

async fn send_block(
    write: &Arc<Mutex<tokio::net::tcp::OwnedWriteHalf>>,
    header: &str,
    kvs: &[(String, String)],
) -> Result<()> {
    let mut payload = String::new();
    payload.push_str(header);
    payload.push_str(":\n");
    for (k, v) in kvs {
        payload.push_str(k);
        payload.push_str(": ");
        payload.push_str(v);
        payload.push('\n');
    }
    payload.push('\n');
    let mut w = write.lock().await;
    w.write_all(payload.as_bytes()).await?;
    Ok(())
}

async fn send_lines(
    write: &Arc<Mutex<tokio::net::tcp::OwnedWriteHalf>>,
    lines: &[&str],
) -> Result<()> {
    let payload = lines.join("\n") + "\n";
    let mut w = write.lock().await;
    w.write_all(payload.as_bytes()).await?;
    Ok(())
}

async fn bind_with_walk(start_port: u16) -> Result<(u16, TcpListener)> {
    let mut last_err: Option<std::io::Error> = None;
    for offset in 0..10 {
        let port = start_port + offset;
        let addr: SocketAddr = ([0, 0, 0, 0], port).into();
        match TcpListener::bind(addr).await {
            Ok(listener) => return Ok((port, listener)),
            Err(err) => last_err = Some(err),
        }
    }
    Err(anyhow!(
        "could not bind BMD protocol on TCP {start_port}-{}: {}",
        start_port + 9,
        last_err
            .map(|e| e.to_string())
            .unwrap_or_else(|| "no error".into())
    ))
}
